### ISSUE TITLE
KoelLabs’ IPA output ≠ eng_to_ipa output

### ISSUE THREAD

**Issue by thuongvovan** [KoelLabs/server#1]
Great work 🎉 — your **xlsr-timit-b0** model produces very clean speech-to-IPA output and is a valuable addition to the community.

## Summary

I’d like to benchmark the model’s phoneme string against a reference IPA produced from text:
```python
import eng_to_ipa as ipa
ref = ipa.convert(
    "The quick brown fox jumped over the lazy dog. "
    "The quick yellow fox jumped over the mad dog."
)
```

eng_to_ipa keeps full IPA table
Your model, trained on the standard 39-phone TIMIT fold, drops stress/closures and collapses some phones of what the problem being solved is. Ex. I'm always frustrated when... -->

## Describe alternatives you've considered

Is there a simple mapping (or script) that turns full IPA from `eng_to_ipa` into the exact 39-phone inventory used in your `vocab.json`?

**Comment by thuongvovan** [comment]
Ah, I also greatly appreciate the Weighted Phonemic Edit Distance (WPED) evaluation method your team proposed. I agree that it provides a more accurate assessment compared to the traditional Phoneme Error Rate (PER).

**Comment by SanderGi** [comment]
Based on your git history, I’m assuming you’re trying to do pronunciation correction and need a “ground truth” to compare the student’s pronunciation to. Definitely be careful with this approach since there are many different dialects and variations in pronunciation that are equally valid and defining just one as correct based on a G2P model/phonemizer risks erasing the other ways of pronunciation. 

It would be better if you had student’s match some audio based ground truth and then ran the model on both the ground truth and student audios to get ground truth and student transcriptions respectively. Then the vocabs would be compatible automatically as well!

If a G2P ground truth really is the best solution for your specific use case, you can use the panphon weighted features to map your G2P model vocab to the closest tokens in our model vocab. This could be done once to precompute a mapping.

**Comment by thuongvovan** [comment]
Thank you for your very accurate advice. I had also considered using an audio-based ground truth, but I encountered difficulties in aligning words with their corresponding phonemes. This is because the model typically generates a continuous sequence of phonemes without boundaries between words. Without a proper mapping between the ground truth text and the ground truth IPA, it becomes challenging to evaluate pronunciation accuracy on a word-by-word basis.

Could you share how you prepared the IPA labels for your data before training your model (whether using a G2P model or a phonemizer)? I’d like to understand your process so I can replicate the IPA labeling and generate suitable ground truth data for my pronunciation assessment system.

**Comment by SanderGi** [comment]
You can align words to phonemes by combining G2P with alignment algorithms like DTW or Needleman-Wunsch. Check out this repo for code. Essentially use G2P to create IPA for each word, then align the actual model compatible vocab IPA ground truth to the G2P to get the word alignment like in this notebook: https://github.com/KoelLabs/ML/blob/main/notebooks/Feedback_Test.ipynb. You can also look into forced alignment algorithms or even use the model timestamp outputs (they are quite accurate due to the CTC loss) and compare against word level timestamps.

We used no G2P when fine-tuning our models (although they were pretrained with G2P data). We used human annotated audio. You can fond audio processing and cleaning code in our ML repo: https://github.com/KoelLabs/ML/tree/main/notebooks/EDA. However, no phoneme to word alignment was necessary for training so our process here is probably not relevant to your use case.

**Comment by SanderGi** [comment]
Sounds good, best of luck with your experiments! Please do update us with what you learn and don’t hesitate to reach out to discuss ideas as you come across them

**Comment by SanderGi** [comment]
The Facebook espeak models are also really great in my experience. From [our evaluation](https://github.com/KoelLabs/ML/blob/main/notebooks/Eval/IPA_eval.ipynb) on a small high quality human annotated test set with 15 hours of speech, they start to struggle on non-native English speech (7% FER) precisely because training on G2P ground truths erase all the accent and pronunciation nuances of non-native speech. This is why we finetuned one of the facebook models (3% FER):

![](https://cdn-uploads.huggingface.co/production/uploads/61dd07bafdc070745eed96fd/vH1zTurznL1qF11E_XE4w.png)

To remove details and obtain a more compact IPA set, you can start simple by using [ipapy](https://pypi.org/project/ipapy/)’s filter_chars to remove diacritics. You’ll then want to take allophones into consideration and map to a symbolset like ARPABet (or the TIMIT variant) that is made specifically for native English.

For our use case of non-native speech, we’ve found that our XLSR-english-01 finetune strikes a good balance between a more broad phoneme set while still capturing a good amount of nuances present in non-native speech. We’ve also found that properly modeling accents is a big improvement over models only trained on G2P (of course it remains to do some experiments with G2P models trained to capture accent variability).
